### PR TITLE
ROR Integration - OAI PMH & Orcid

### PR DESCRIPTION
**Pull request by vins01-4science**
**Please note that this development has been funded by the University of California - California Digital Library.**

## References
* This PR builds on top of DSpace/Dspace#9238
* Related to Dspace/dspace-angular#2719

## Description
This feature introduces ROI metadata visualization inside the OAI-PMH and its synchronization with ORCID profile, if configured any.

## Instructions for Reviewers
First of all, you need to enable the [OAI-PMH](https://wiki.lyrasis.org/display/DSDOC7x/OAI) and [ORCID synchronization](https://wiki.lyrasis.org/display/DSDOC7x/ORCID+Integration#ORCIDIntegration-Overview) feature. 

Configuration steps:
1. Enable the OAI server inside oai.cfg:
```
2. Enable (true) or disable (false) OAI-PMH server
oai.enabled = true
# When enabled, OAI-PMH server is available at this path
oai.path = oai
```
3. Configure some collections to use the standard OpenAIRE form support inside item-submission.xml:
```xml
<name-map collection-handle="123456789/2" submission-name="openAIREPublicationSubmission" />
<name-map collection-handle="123456789/3" submission-name="openAIREPersonSubmission" />
<name-map collection-handle="123456789/5" submission-name="openAIREProjectSubmission" />
<name-map collection-handle="123456789/4" submission-name="openAIREOrganizationSubmission" />
```
4. Initialize all the relationships and metadata related to the openaire integration:
```
[/dspace]/bin/dspace initialize-entities -f [/dspace]/config/entities/openaire4-relationships.xml
[/dspace]/bin/dspace registry-loader -metadata [/dspace]/config/registries/openaire4-types.xml
```
5. Enable OAI interface:
```
oai.enabled = true
```
6. Replace the file `[/dspace/]config/spring/api/virtual-metadata.xml` with the one named `[/dspace/]config/spring/api/virtual-metadata.xml.openaire` by renaming this one, in order to see all the additional openaire information inside the OAI-PMH interface.
7. Just run the command:
```
[/dspace]/bin/dspace oai import -c
```
To reinitialize the oai solr core with your configuration.

1. Configure orcid API integration: 
```
# These URLs are for testing against ORCID's Sandbox API
# These are only useful for testing, and you must first request a Sandbox API Key from ORCID
orcid.domain-url= https://sandbox.orcid.org
orcid.api-url = https://api.sandbox.orcid.org/v3.0
orcid.public-url = https://pub.sandbox.orcid.org/v3.0
# Keep in mind, these API keys MUST be for the Sandbox API if you use "sandbox.orcid.org" URLs above!
orcid.application-client-id = <YOUR-SANDBOX-ORCID-CLIENT-ID>
orcid.application-client-secret = <YOUR-SANDBOX-ORCID-CLIENT-SECRET>
Enable orcid login feature - plugin.sequence.org.dspace.authenticate.AuthenticationMethod = org.dspace.authenticate.OrcidAuthentication
```
2. Enable orcid synchronization
```
orcid.synchronization-enabled = true
```
3. Enable orcid synchronization consumer (orcidqueue)
```
event.dispatcher.default.consumers = versioning, discovery, eperson, orcidqueue
```
4. Enable orcid validation:
```
orcid.validation.work.enabled = true
orcid.validation.funding.enabled = true
```
5. Configure new ROR identifier mapping:
```
orcid.mapping.organization.identifiers = organization.identifier.ror::ROR
orcid.validation.organization.identifier-sources = ROR
```

List of changes in this PR:
* Changes coming from the live-import feature of ROR
* Changes involving configuration files to enable the new features

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

### OAI Integration
| Step  | Screenshot |
| ------- | ---------------- |
|Import an `Organization` using the ROR live import feature| image |
| Link a `Project` to an `Organization` entity | image |
| Create a `Publication` with the OpenAIRE form configuration and link it to the `Project`| image |
| Update the OAI index | `[/dspace]/bin/dspace oai import -c`|
| Visualize the entity inside the `oai` interface (`[dspace-url]/server/oai/...`) | image|

### ORCID Synchronization
To verify the OAI integration, you need to:
| Step | Screenshot |
| ------ | ----------------- |
| Login with an ORCID user that has a Researcher Profile and that is able to submit |image |
| Enable ORCID Synchronization from its profile page | image|
| Import a ROR `Organization` with the live import feature | image|
| Link a `Project` to the `Organization` and the ORCID User | image|
| You should have a new `Project` entry inside the ORCID Synchronization Queue of your profile | image |
| Synchronize the change to ORCID, and you'll see the new Funding with the related ROR identifier | image image|

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

**Review by tdonohue — COMMENTED** [review]
@vins01-4science : This PR appears to be build on top of #9237.  So, I noticed most of the code is the same.  

In code reviewing this PR, I've only reviewed the diff between these branches: https://github.com/4Science/DSpace/compare/main_CST-12825...4Science:DSpace:main_CST-12826

All in all, this appears to be a tiny PR that is only made up of configuration changes (primarily to OAI-PMH and `virtual-metadata.xml.openaire4`).   This seems like a very obvious PR to merge quickly once #9237 is approved/merged. 

I'll do a more thorough test of this once #9237 is ready to go... but the code changes here look reasonable/correct.

File: dspace/config/submission-forms.xml
Comment (paulo-graca) [inline dspace/config/submission-forms.xml:541]: I don't understand this change. Why, in the context of a project, do you consider a funder as a dc.contributor.other?

If you check https://schema.org/ResearchProject you will find the funder there. So this change doesn't make sense to me.

**Review by paulo-graca — CHANGES_REQUESTED** [review]
Thank you @vins01-4science for this contribution. ROR integration is a very welcomed feature. I've just add some comments related with missing java docs and some schema approaches. I didn't test it, but, overall, the code contributions looks fine to me.

File: dspace/config/submission-forms.xml
Comment (vins01-4science) [reply dspace/config/submission-forms.xml:541]: Is not a funder, it's the linked orgunit in the context of the project.
This relationship is mapped by the type `isOrgUnitOfProject` you can find it also inside the `virtual-metadata.xml` configuration, here:  https://github.com/DSpace/DSpace/blob/main/dspace/config/spring/api/virtual-metadata.xml#L100-L102.

**Comment by tdonohue** [comment]
@vins01-4science : Can you rebase this on the latest `main` now that #9238 has been merged?

**Comment by vins01-4science** [comment]
> @vins01-4science : Can you rebase this on the latest `main` now that #9238 has been merged?

@tdonohue , this branch have been rebased correctly, now we should wait for @paulo-graca review on my last comment.

Thank you.

**Review by paulo-graca — APPROVED** [review]
Thank you so much @vins01-4science !

**Review by tdonohue — APPROVED** [review]
👍 Thanks @vins01-4science !  This looks good to me too.  Gave it some simple testing and appears to be working.

**Pull request merged**
